tools/testbench/topology_ipc4.c

```diff
@@ -330,10 +331,84 @@ int tb_match_audio_format(struct testbench_prm *tp, struct tplg_comp_info *comp_
 	return 0;
 }
 
+static int tb_set_pin_formats(struct tplg_comp_info *comp_info,
+			      struct ipc4_base_module_cfg *base_cfg,
+			      struct ipc4_base_module_cfg_ext *base_cfg_ext, int pin_type)
+{
+	struct sof_ipc4_pin_format *formats_array;
+	struct sof_ipc4_pin_format *pin_format_item;
+	struct sof_ipc4_pin_format *pin_format;
+	struct sof_ipc4_pin_format *base_cfg_ext_pin_formats =
+		(struct sof_ipc4_pin_format *)base_cfg_ext->pin_formats;
+	int format_list_count;
+	int num_pins;
+	int i, j;
+	int pin_format_offset = 0;
+
+	if (pin_type == 0) {
+		num_pins = base_cfg_ext->nb_input_pins;
+		formats_array = comp_info->available_fmt.input_pin_fmts;
+		format_list_count = comp_info->available_fmt.num_input_formats;
+	} else {
+		num_pins = base_cfg_ext->nb_output_pins;
+		pin_format_offset = base_cfg_ext->nb_input_pins;
+		formats_array = comp_info->available_fmt.output_pin_fmts;
+		format_list_count = comp_info->available_fmt.num_output_formats;
+	}
+
+	for (i = pin_format_offset; i < num_pins + pin_format_offset; i++) {
+		pin_format = &base_cfg_ext_pin_formats[i];
+
+		if (i == pin_format_offset) {
+			if (pin_type == 0) {
+				pin_format->buffer_size = base_cfg->ibs;
+				/* Note: Copy "struct ipc4_audio_format" to
+				 * "struct sof_ipc4_pin_format". They are same
+				 * but separate definitions. The last member fmt_cfg
+				 * is defined with bitfields for channels_count, valid_bit_depth,
+				 * s_type, reservedin ipc4_audio_format.
+				 */
+				memcpy(&pin_format->audio_fmt, &base_cfg->audio_fmt,
+				       sizeof(pin_format->audio_fmt));
+			} else {
+				pin_format->buffer_size = base_cfg->obs;
+				/* TODO: Using workaround, need to find out how to do this. This
+				 * is copied kernel function sof_ipc4_process_set_pin_formats()
+				 * from process->output_format. It's set in
+				 * sof_ipc4_prepare_process_module().
+				 */
+				memcpy(&pin_format->audio_fmt, &base_cfg->audio_fmt,
+				       sizeof(pin_format->audio_fmt));
```

Comment:
Note: After talking with @ranj063 I now understand resolving this properly need another prepare() pipelines walk step that does not yet exist in testbench and plugin. I'm proposing to keep this as-is for now and wait for development in plugin where things are more constrained by multiple processes. This workaround works for most components where first pin format for input and output is the same.